marlowe/src/Language/Marlowe/Semantics.hs

```diff
@@ -473,6 +475,18 @@ evalValue env state value = let
         AddValue lhs rhs     -> eval lhs + eval rhs
         SubValue lhs rhs     -> eval lhs - eval rhs
         MulValue lhs rhs     -> eval lhs * eval rhs
+        DivValue lhs rhs     -> let n = eval lhs
+                                in if n == 0 then 0 else let
+                                    d = eval rhs
+                                in if d == 0 then 0 else let
+                                    (q, r) = n `quotRem` d
+                                    ar = abs r * 2
+                                    ad = abs d
+                                in if ar < ad then q -- reminder < 1/2
+                                   else if ar > ad then q + signum n * signum d -- reminder > 1/2
+                                   else let -- reminder == 1/2
+                                even = q `Builtins.remainderInteger` 2 == 0
+                                in if even then q else q + signum n * signum d
```

Comment:
```suggestion
                                   d = eval rhs
                                   (q, r) = n `quotRem` d
                                   ar = abs r * 2
                                   ad = abs d
                                   even = q `Builtins.remainderInteger` 2 == 0
                                in if d == 0 then 0 else
                                   if ar > ad || (ar == ad && qIsOdd)
                                   then q + signum q
                                   else q
```
I have 4 suggestions, so I've packed them in one comment to be able to make a code suggestion:
- I think the initial `n == 0` check doesn't make any difference in the result but I am not sure
- I think `signum n * signum d` be the same as `signum q`, so we may use that
- `even` seems to shadow a `Prelude` function, so I would call it `qIsEven` or something like that. In the end, it seems `qIsOdd` would be simpler in this case.
- I have found a way to collapse several of the `if` together, mainly I am doing this because it will probably help symbolic execution complexity to minimise the number of `ite`

Reply:
The reason I do `let`s inside `if`s is to avoid Plutus evaluations on-chain, because those are strict unlike Haskell. If `n` is 0 we even not try to compute `d`, and if case of `ar <> ad` we do not compute `qIsOdd`.
It's fine for FSSemantics though, will change it there.

Reply:
Although, it's not a huge problem considering the size of Marlowe validator :)

Reply:
Ah, I see, yes, good point 🤔, it may produce an exception if we do that actually. So, another idea, maybe we can shortcut in the `d` instead, and have something like this, precomputing `quotRem` and such is not a lot of waste, probably more efficient than branching in practice:
```suggestion
                                   d = eval rhs
                                in if d == 0
                                   then 0
                                   else let n = eval lhs
                                        (q, r) = n `quotRem` d
                                        ar = abs r * 2
                                        ad = abs d
                                        qIsOdd = q `Builtins.remainderInteger` 2 /= 0
                                        in if n /= 0 && (ar > ad || ar == ad && qIsOdd)
                                           then q + signum n * signum d
                                           else q
```
We would have to remove the first computation of `n` too

Reply:
How it can produce an exception? I checked and branching is a very cheap builtin in Plutus Core unlike quot/rem functions. How about this?
```
                                let n = eval lhs
                                in if n == 0 then 0 else let
                                    d = eval rhs
                                in if d == 0 then 0 else let
                                    (q, r) = n `quotRem` d
                                    ar = abs r * 2
                                    ad = abs d
                                in if ar < ad then q else let
                                    qIsOdd = q `Builtins.remainderInteger` 2 /= 0
                                in if qIsOdd || ar > ad then q + signum n * signum d else q
```

Reply:
I checked the size of Marlowe validator for different approaches, and it looks like GHC does a good job of optimising ASTs and dealing with repeating `q + signum n * signum`. I see no size differences. I'll go with my original code and rename `even` to `qIsEven`